### PR TITLE
Handle generic bounds in a uniform way in HIR

### PR DESCRIPTION
**Pull request by cjgillot**
Generic bounds in HIR used to be split between bounds in the parameter definition and bounds in a where clause. This PR attempts to store all of those as where predicates. 

This effectively desugars
```rust
fn foo<T: Default, U>(x: impl Copy) where U: Clone
```
into
```rust
fn foo<T, U, _V>(x: _V) where T: Default, U: Clone, _V: Copy
```
(where _V is actually hidden and called "impl Copy").

I managed to make compiler warnings more uniform.
About rustdoc: is making this desugaring user-visible acceptable?
About clippy: I don't understand the subtle logic in the `needless-lifetimes` lint.

r? @estebank

### CODE REVIEW

**Comment by CraftSpider** [comment]
I'm not sure I'd mind seeing all bounds as `where`, but I'd definitely defer to other rustdoc members on that. If this is going to merge though, please ensure you also bump `FORMAT_VERSION` in the rustdoc-json-types module. It needs to be increased on every breaking change.

**Comment by aDotInTheVoid** [comment]
I think its important to show that their is a difference in rustdoc html, as `impl Trait` cannot be used with the turbofish, although AFAIKT the pr still does that.

[The JSON output also has a significant regression](https://gist.github.com/aDotInTheVoid/8685b3b4e8b526d7e0c0c6515d0e55a6#file-diff-diff) in that `pub fn by_impl_trait(x: impl Foo) {}` is given arguments `["x", {"inner": "impl Foo", "kind": "generic"}]`, and no mension of the `id` of `Foo` anywhere.

Unrelated, but we should probably merge [`src/etc/check_missing_items.py`](src/etc/check_missing_items.py) into jsondocck

**Comment by camelid** [comment]
`impl Trait` should definitely **_not_** be shown as a generic.

**Comment by cjgillot** [comment]
@aDotInTheVoid how do I run the json tests?

File: src/test/ui/associated-type-bounds/type-alias.stderr
Comment (estebank) [inline src/test/ui/associated-type-bounds/type-alias.stderr:78]: This change isn't a big problem, but it would be nice not to have this change.

File: src/test/ui/generic-associated-types/issue-86483.stderr
Comment (estebank) [inline src/test/ui/generic-associated-types/issue-86483.stderr:10]: The previous suggestions were already "wrong", but the new output is "wrong". I wonder if we should "just" detect these and remove the suggestion.

File: src/test/ui/impl-trait/must_outlive_least_region_or_bound.stderr
Comment (estebank) [inline src/test/ui/impl-trait/must_outlive_least_region_or_bound.stderr:116]: These would benefit from being `span_suggestion_verbose` 🤔

**Comment by Urgau** [comment]
> @aDotInTheVoid how do I run the json tests?

`./x.py test src/test/rustdoc-json`

**Comment by CraftSpider** [comment]
> how do I run the json tests?

`x.py test src/test/rustdoc-json --stage 2` is the command I run

File: src/test/ui/suggestions/suggest-impl-trait-lifetime.stderr
Comment (estebank) [inline src/test/ui/suggestions/suggest-impl-trait-lifetime.stderr:5]: Another case that would be helped by `span_suggestion_verbose`

**Review by estebank — COMMENTED** [review]
I need to go over the code changes, but the output changes look reasonable to me (with some nitpicking for things we could improve) so it's likely that the code changes themselves are too. I'll read the changes now.

**Comment by estebank** [comment]
I agree with @camelid, if the signature uses `impl Trait`, then the docs and error messages need to refer to `impl Trait` and not `X: Trait` in all cases. The internal representation can change so that they are handled more gracefully though.

**Comment by flip1995** [comment]
> About clippy: I don't understand the subtle logic in the `needless-lifetimes` lint.

I don't see any changes to Clippy's tests. The changes to lifetimes.rs seem reasonable AFAICT. Do you get test failures in Clippy?

**Comment by cjgillot** [comment]
> I agree with @camelid, if the signature uses `impl Trait`, then the docs and error messages need to refer to `impl Trait` and not `X: Trait` in all cases. The internal representation can change so that they are handled more gracefully though.

Of course, the desugaring will not appear in docs and errors. Only the internal representation is changed by this PR.

File: src/tools/clippy/tests/ui/extra_unused_lifetimes.stderr
Comment (cjgillot) [inline src/tools/clippy/tests/ui/extra_unused_lifetimes.stderr:9]: @flip1995 I mean this change, and the one below. I don't understand how clippy use to account for lifetime uses in bounds and not where clauses.

Comment (flip1995) [reply src/tools/clippy/tests/ui/extra_unused_lifetimes.stderr:9]: I can't really tell what would cause this from just looking at the Clippy changes. I'll have to review and understand the other changes to rustc in this PR. This might take me a few days until I get to that though.

**Comment by estebank** [comment]
It seems like it hits diesel in particular 😮

**Comment by cjgillot** [comment]
Perf improvement balance perf regressions.

**Comment by cjgillot** [comment]
@estebank Are there still any unaddressed reviews?
I'll rebase after next review round, the test changes are pretty bitrotty.

**Comment by flip1995** [comment]
From the Clippy side: This introduces 2 FNs. I'm not sure how those could be avoided. I would recommend that you open an issue in the Clippy repo about them after this PR is merged.

File: compiler/rustc_hir/src/hir.rs
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:573]: That comment does not match what the function is doing

File: compiler/rustc_resolve/src/late/lifetimes.rs
Comment (oli-obk) [inline compiler/rustc_resolve/src/late/lifetimes.rs:1421]: Add a FIXME so we'll try to use the span of the definition of the lifetime instead of the use

**Comment by oli-obk** [comment]
r=me after nits addressed and a rebase

**Comment by GuillaumeGomez** [comment]
Please don't merge until all changes in rustdoc test files have been reverted. They're not supposed to be updated considering they ensure it's working as expected (unless it's a rebase issue, then please ignore my comment).

File: src/rustdoc-json-types/lib.rs
Comment (GuillaumeGomez) [inline src/rustdoc-json-types/lib.rs:349]: Why this change? You're removing a lot of information.

**Comment by oli-obk** [comment]
Note that the rustdoc changes are because there are no more bounds on generics, but all of them have been converted to where bounds. There has been no loss of information as far as I can tell.

Maybe some before/after screenshots of the changes could help?

**Comment by GuillaumeGomez** [comment]
I was just looking at the rustdoc test changes. But you're right, most changes are not a problem. However, the search results changes are worrying me.

File: src/rustdoc-json-types/lib.rs
Comment (GuillaumeGomez) [reply src/rustdoc-json-types/lib.rs:349]: Nevermind, just went through the rustdoc tests again and it seems to be working as expected.

File: src/test/rustdoc/issue-20727-4.rs
Comment (Urgau) [inline src/test/rustdoc/issue-20727-4.rs:20]: This is unfortunate, I understand that it means the same but the source not being anymore rendered the same way as it was written isn't great. If I wasn't aware of this change I would considered this a being a serious regression even-trough having the "where" clause means the same thing.

Comment (GuillaumeGomez) [reply src/test/rustdoc/issue-20727-4.rs:20]: I opened a discussion about it on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/converting.20bounds.20into.20where.20bounds.3F).

**Comment by cjgillot** [comment]
This latest version removes the changes to rustdoc output, to allow progress on that PR independently of the team's decision.

**Review by oli-obk — APPROVED** [review]
r=me with CI passing

File: compiler/rustc_resolve/src/late/lifetimes.rs
Comment (cjgillot) [reply compiler/rustc_resolve/src/late/lifetimes.rs:1417]: Commit 0b8a80b3e9db1181eebaf246df0f634a64179083 does that. I'm not convinced by the diagnostic: the span is a bit far from the reason it's generated.

Comment (oli-obk) [reply compiler/rustc_resolve/src/late/lifetimes.rs:1417]: OK yea, that's worse. Maybe we'll need to create a multi span diagnostic or just a suggestion that removes the param and renames all uses to 'static.

Anyway, off-topic

**Comment by oli-obk** [comment]
r=me after rebase and ui tests blessed

**Comment by Mark-Simulacrum** [comment]
@cjgillot It looks like the pre-merge perf justification in https://github.com/rust-lang/rust/pull/93803#issuecomment-1060028122 was that the improvements balanced regressions, but post-merge we are not seeing any improvements. I think the (presumably) improve code quality/maintenance may still justify this change, but it would be good to re-assess from your perspective.

**Comment by jackh726** [comment]
I must have missed this or forgotten about it.

This seems counterintuitive to me in the context of the TAIT refactoring. Of course, @oli-obk knows more than me on this, but surprising nonetheless.

**Pull request merged**
